analyzer/src/main/kotlin/managers/Yarn2.kt

```diff
@@ -0,0 +1,141 @@
+     * for the current project e.g. yarnPath: .yarn/releases/yarn-3.2.1.cjs.
+     * This map holds the mapping between the directory and their Yarn2 executables.
+     */
+    private val yarn2CommandsByPath: MutableMap<File, String> = mutableMapOf()
+
+    /**
+     * A NPM package manager, used to call NPM when Yarn2+ does not provide enough information.
+     */
+    private val delegateNpmAnalyzer = Npm("npm", analysisRoot, analyzerConfig, repoConfig)
+
+    companion object {
+        const val YARN2_RESOURCE_FILE = ".yarnrc.yml"
+
+        // The property in `.yarnrc.yml`containing the path to the Yarn2+ executable.
+        const val YARN_PATH_PROPERTY_NAME = "yarnPath"
+    }
+
+    override fun hasLockFile(projectDir: File) = hasYarnLockFile(projectDir)
+
+    override fun command(workingDir: File?): String {
+        return workingDir?.let {
+            yarn2CommandsByPath.getOrPut(it) {
+                val yarnConfig = yamlMapper.readTree(workingDir.resolve(YARN2_RESOURCE_FILE))
+                val yarnCommand = yarnConfig.get(YARN_PATH_PROPERTY_NAME)
+                yarnCommand?.let { command ->
+                    val yarnExecutable = workingDir.resolve(command.textValue())
+
+                    // TODO: Yarn2 executable is a `cjs` file. Check if under Windows it needs to be run with `node`.
+
+                    // TODO: This is a security risk to blindly run code coming from a repository other than ORT's. ORT
+                    //  should download the Yarn2 binary from the official repository and run it.
+                    require(yarnExecutable.exists()) {
+                        "The Yarn2+ program '${yarnExecutable.name}' does not exist."
+                    }
+
+                    if (!yarnExecutable.canExecute()) {
+                        log.warn {
+                            "The Yarn2+ program '${yarnExecutable.name}' should be executable. Changing its rights."
+                        }
+                        require(yarnExecutable.setExecutable(true)) {
+                            "Cannot set the Yarn2+ program to be executable."
+                        }
+                    }
+                    yarnExecutable.absolutePath
+                } ?: throw IllegalStateException("No Yarn2+ executable could be found in 'yarnrc.yml'.")
+            }
+        } ?: "yarn2+ executable' in .yarn/releases/ and configured in .yarnrc.yml"
+    }
+
+    override fun getVersion(workingDir: File?): String =
+        // `getVersion` with a `null` parameter is called once by the Analyzer to get the version of the global tool.
+        // For Yarn2+, the version is specific to each definition file being scanned therefore a global version doesn't
+        // apply.
+        // TODO: An alternative would be to collate the versions of all tools in `yarn2CommandsByPath`.
+        if (workingDir == null) "N/A" else super.getVersion(workingDir)
```

Comment:
The convention is to return an empty string if the version is unknown / cannot be determined.